### PR TITLE
added hashfile-function to lib/crypto

### PR DESCRIPTION
**Pull request by ineiti**
Added a HashFile function to the `lib/crypto` that always returns 32 bytes.

### CODE REVIEW

**Comment by nikkolasg** [comment]
A remark coming from Ismail's remark: we should start handling everything more as a "stream" rather than a simple bulk of bytes. For example here, we could redesign using a `io.Reader` as argument and reading chunk by chunk so the whole file does not have to get into memory.

**Comment by Liamsi** [comment]
Any idea why travis is failing here? 
I realized that we already hash correctly in the cli branch (must have been outdated output we saw on Thursday). Furthermore, it could be a better approach to simply have an public `func Hash(b []byte) ([]byte, error)` (can be used for the files, too) somewhere instead of `HashFile`?

**Comment by ineiti** [comment]
If you want to have a `Hash`-function, you won't be able to do 'streaming' on a file.

**Comment by nikkolasg** [comment]
Something like:
```
// do the *real* hashing
func Hash(b []byte) []byte {
   ...
}
func HashStream(stream io.Reader) []byte {
     for {
        // Something like...
        b := stream.Read()
        Hash(b)
    }
}
func HashFile(filename string) []byte {
   f := file.Open(filename)
   return HashStream(f)
}
```
Avoiding the brittleness in our code ... !

**Comment by ineiti** [comment]
But how do you keep the state of the hash during multiple calls to `Hash`?

**Comment by nikkolasg** [comment]
When you have a `Hash` object, calling `Sum(nil)` returns the current state <=> hash of the input, but it does not change the *state* of the Hash. You can still input something to the hash after.
```
        // Sum appends the current hash to b and returns the resulting slice.
        // It does not change the underlying hash state.
        Sum(b []byte) []byte
```
Source: https://golang.org/pkg/hash/

**Comment by ineiti** [comment]
And how do you handle two goroutines hashing their file?

**Comment by nikkolasg** [comment]
See updated snippet code.

**Comment by ineiti** [comment]
Is this worth being pursued or do we just close it?

**Comment by Liamsi** [comment]
> Is this worth being pursued or do we just close it?

I think what Nicaols proposed makes sense (and we will need sth. like in different places anyways). But I'm also OK to just close it, if you both agree.

**Comment by ineiti** [comment]
Can you just add an issue in `Features`, please? Then close and delete this branch.

**Comment by nikkolasg** [comment]
Just done it, tell me if that seems ok otherwise we rollback and merge/close.

File: lib/crypto/hash.go [hunk 1]
```diff
@@ -4,34 +4,69 @@ import (
...
+// HashStream returns the hash of the stream reading from it chunk by chunk of
+// size DefaultChunkSize
+func HashStream(suite abstract.Suite, stream io.Reader) ([]byte, error) {
+	return hashStream(suite, stream, DefaultChunkSize)
+}
+
+// Hash a stream using chunks of size
+func HashStreamChunk(suite abstract.Suite, stream io.Reader, chunkSize int) ([]byte, error) {
+	if chunkSize < 1 {
+		return nil, errors.New("Wront chunksize value")
```
Comment (ineiti) [inline lib/crypto/hash.go]: Typo - "wrong"

File: lib/crypto/hash.go [hunk 2]
```diff
@@ -4,34 +4,69 @@ import (
...
 	return hash.Sum(nil), nil
 }
 
+// HashStream returns the hash of the stream reading from it chunk by chunk of
+// size DefaultChunkSize
+func HashStream(suite abstract.Suite, stream io.Reader) ([]byte, error) {
+	return hashStream(suite, stream, DefaultChunkSize)
+}
+
+// Hash a stream using chunks of size
```
Comment (ineiti) [inline lib/crypto/hash.go]: Docu-error: first word has to be the name of the function
enumerate arguments

**Comment by Liamsi** [comment]
For me it would be OK to merge it

File: lib/crypto/hash.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+)
+
+// Hash simply returns the Hash of the slice of bytes given
+func Hash(hash h.Hash, buff []byte) ([]byte, error) {
+	_, err := hash.Write(buff)
+	return hash.Sum(nil), err
+}
+
+// Size of the chunk  used to read a stream
+const DefaultChunkSize = 1024
```
Comment (ineiti) [inline lib/crypto/hash.go]:
```hash.go:18:1: comment on exported const DefaultChunkSize should be of the form "DefaultChunkSize ..."
```

Comment (nikkolasg) [inline lib/crypto/hash.go]: Damn. Also forgot to put the lint verification. Done. Thx.

**Pull request merged**